internal/pkg/template/env.go

```diff
@@ -0,0 +1,74 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package template
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// Paths of service cloudformation templates under templates/services/.
+const (
+	// EnvCFTemplatePath is the path where the cloudformation for the environment is written.
+	EnvCFTemplatePath     = "environment/cf.yml"
+	envNestCFTemplatePath = "environment/nest/cf/%s.yml"
+)
+
+var (
+	// Template names under "environment/nest/cf/".
+	nestEnvCFTemplateNames = []string{
+		"cfn-execution-role",
+		"custom-resources",
+		"custom-resources-role",
+		"environment-manager-role",
+		"lambda",
+		"vpc-resources",
+	}
+)
+
+// EnvOpts holds data that can be provided to enable features in an environment stack template.
+type EnvOpts struct {
+	DNSDelegationLambda       string
+	ACMValidationLambda       string
+	EnableLongARNFormatLambda string
+	ImportVpc                 *ImportVpcOpts
+	VpcConfig                 *AdjustVpcOpts
+}
+
+// ImportVpcOpts holds the fields to import VPC resources.
+type ImportVpcOpts struct {
+	ID               string // ID for the VPC.
+	PublicSubnetIDs  []string
+	PrivateSubnetIDs []string
+}
+
+// AdjustVpcOpts holds the fields to adjust default VPC resources.
+type AdjustVpcOpts struct {
+	CIDR               string // CIDR range for the VPC.
+	PublicSubnetCIDRs  []string
+	PrivateSubnetCIDRs []string
+}
+
+// ParseEnv parses an environment's CloudFormation template with the specified data object and returns its content.
+func (t *Template) ParseEnv(data interface{}, options ...ParseOption) (*Content, error) {
```

Comment:
If you look at these `Parse*` methods, we decided on purpose to take a specific struct so that we are guaranteed not to run into a runtime issue by forcing the data passed in to always render a valid template. For example:
https://github.com/aws/copilot-cli/blob/7cb56c0cfee1ddc550c5d725eee7687a24dee6b8/internal/pkg/template/service.go#L64

I think I made a mistake with having these defined as methods, instead what do you think of defining them as functions in this package. Here is a sample interface: https://gist.github.com/efekarakus/3edbd280587c8433b0811307c2027b6a with an explanation at the bottom.

Reply:
Let me know your thoughts with the API above. 
If you like it, we can keep the current PR as is but do have the next PR be a refactor of the template pkg.

Reply:
yeah since this pr is already big enough. I'll do the refactoring in the next one!